src/core/lib/transport/byte_stream.cc

```diff
@@ -55,17 +55,20 @@ void SliceBufferByteStream::Orphan() {
 
 bool SliceBufferByteStream::Next(size_t max_size_hint,
                                  grpc_closure* on_complete) {
-  GPR_ASSERT(cursor_ < backing_buffer_.count);
+  GPR_DEBUG_ASSERT(backing_buffer_.count > 0);
   return true;
 }
 
 grpc_error* SliceBufferByteStream::Pull(grpc_slice* slice) {
-  if (shutdown_error_ != GRPC_ERROR_NONE) {
+  if (GPR_UNLIKELY(shutdown_error_ != GRPC_ERROR_NONE)) {
     return GRPC_ERROR_REF(shutdown_error_);
   }
-  GPR_ASSERT(cursor_ < backing_buffer_.count);
-  *slice = grpc_slice_ref_internal(backing_buffer_.slices[cursor_]);
-  ++cursor_;
+
+  GPR_DEBUG_ASSERT(backing_buffer_.count > 0);
```

Comment:
Why not using `grpc_slice_buffer_take_first()` directly?

Reply:
I specifically avoided that to save a GPR_ASSERT branch (and the call/ret). grpc_slice_buffer_take_first() has to sanitize its inputs, so to speak, since it is marked GPRAPI.

Reply:
I don't think `GPR_ASSERT` is a necessity there in the GPRAPI.  Personally, I think that can be made `GPR_DEBUG_ASSERT` everywhere in the hot path. @vjpai  is the authority though.

Either way, I still think we shouldn't dup the code and we should `grpc_slice_buffer_take_first()` here.

Reply:
+1 on not duplicating the code.

Reply:
Updated.

Reply:
Thank you @arjunroy  for addressing the comment.